core/chain_makers.go

```diff
@@ -282,7 +303,14 @@ func GenerateChain(config *params.ChainConfig, parent *types.Block, engine conse
 		}
 		if b.engine != nil {
 			// Finalize and seal the block
-			block, _ := b.engine.FinalizeAndAssemble(chainreader, b.header, statedb, b.txs, b.uncles, b.receipts)
+			shanghai := config.IsShanghai(new(big.Int).SetUint64(b.header.Time))
+			if shanghai && b.withdrawals == nil {
+				// need to make empty list to denote non-nil, but empty withdrawals to calc withdrawals hash
+				b.withdrawals = make([]*types.Withdrawal, 0)
+			} else if !shanghai && b.withdrawals != nil {
+				panic("withdrawals set before activation")
+			}
+			block, _ := b.engine.FinalizeAndAssemble(chainreader, b.header, statedb, b.txs, b.uncles, b.receipts, b.withdrawals)
```

Comment:
The checks here seem superfluous -- and should instead be placed _inside of_  `FinalizeAndAssemble`, and callers thereof would not need to do the checks explicitly.